readme.txt
README file for 'NTLM Authorization Proxy Server' v.0.9.9

Release Purpose:
----------------
This is a re-release of 0.9.8.11 as 0.9.9 'stable'. See changelog.txt for full description.

There is an issue with APS working as a standalone proxy. It serves requests from
an http-client one by one and allows persistent connections, then it may receive several
request in very sort time to one thread, and one of them may be to almost dead banner
site, then all the requests made after that one will be waiting till that "bad" connection
will be closed due to timeout. So I suggest switching off HTTP/1.1 presistent connections
in your browser when you are using APS for web (not proxy) authentication and surfing
banner rich evironment.

This is a beta version, so I expect it to have bugs. If something goes
wrong, please log a support request on the sourceforge.net project page.

Please read this file to the end, it contains useful installation
instructions and other information.


1. WHAT IS 'NTLM Authorization Proxy Server'?
---------------------------------------------

'NTLM Authorization Proxy Server' is a proxy-like software, that will authorize you
at MS proxy server and at web servers (ISS especially) using MS proprietary NTLM
authorization method and it can change some values in your client's request header
so that those requests will look like ones made by MS IE. It is written in Python
language. See www.python.org.

2. WHY MAY I WANT TO USE 'NTLM Authorization Proxy Server'?
-----------------------------------------------------------
When your web browser uses proxy server to surf the Web from within your local
network your may be required to authenticate youself to use proxy. There are
two methods to do this: using 'Basic' authorization method and 'Digest' method.

'Basic' is very basic;) and not secure, but considered often as sufficient to check
users inside your local network. The second - 'Digest' - is quite secure but, as I
could note, use of it considered as 'too much' in order to protect your own proxy
server from your own people. Well, 'Digest' is far more complicate method and I don't
know any software that can use it, I think it is becuse this method simply 'overkill' for
most cases. 'Basic' method is quite common and is standard method to authenticate
yourself at proxies.

But MicroSoft, being MS, has developed instead of standard methods two its own
authentication methods - 'NTLM' (NT LanManager) and 'Negotiate'. I don't know anything
about the second one, but I spend some time on 'NTLM'.

Because NTLM (NT LanManager) is a proprietary algorithm it is looks like that
only MS's products can use it (IE3+ as far as I know). And it gives you a bunch of
fun things: If proxy server that you have to use is set so it requres your to
authenticate yourself with 'NTLM' algorithm then you will be able to use ONLY MS's
products with this proxy server. No ReGet, no WGET, no Junkbuster, no Netscape
Navigator, no Lynx or your local squid and no lots of other good software.

Why admin of proxy server in your local network may want to ban standard 'Basic'
and require 'NTLM'? I don't know. May be because he likes MS too much;).


3. HISTORY BEHIND THE 'NTLM Authorization Proxy Server'.
--------------------------------------------------------

I got the idea for 'NTLM Authorization Proxy Server' when admin in our network
swiched MS Proxy so that we had to use 'NTLM'. It was simply disaster. Everything
stopped working but IE5.5.

So then I had realized that there was no solution to make my favorite programs work with
it I decided to do something. And here you are.


4. REQUIREMENTS
---------------

You will need a thread-safe Python installation, version 1.5.2 or later.
Python is available from http://www.python.org/ . Get it, install it, and you
are ready to try. Python also comes with all GNU/Linux distributions that
I am aware of, so if you are a GNU/Linux user you should have that in your
box already.

Please note that I have not test it with Python versions higher than 1.5.2.
Therefore you may have problems with later Pythons, but I hope you will not.

5. LICENSING & PRICING
----------------------

'NTLM Authorization Proxy Server' is distributed under the GNU Public License,
which is included in this archive (see file COPYING).

The above mean that 'NTLM Authorization Proxy Server' is pretty much free.
You have to pay nothing for it.

6. CREDITS
----------
Thanks to Nathan Lineback <lineback@pla-netx.com> for reporting about persistent
connection to MS server with multiple authorization requests. I had never heard of
such a behavior before that.

Janek Schwarz <j.schwarz@i-change.de> added command line option '-c' for config files
other than default server.cfg in working directory. Thanks for that to him.

Thanks to Stephen D. Cohen for very useful suggestions and cooperation.

Patch for Python >=1.6 by Markus Indenbirken (markus_i@gmx.de)

The most useful info has been got from "NTLM Authentication Scheme for HTTP"
page by Ronald Tschalar (ronald@innovation.ch). I included the page in the distribution
because it looks like that the server it contains is going down.
(http://www.innovation.ch/java/ntlm.html)

md4.py and des_c.py had been translated from md4.c and des.c from
"the Python Cryptography Toolkit, version 1.0.0 Copyright (C) 1995, A.M. Kuchling"



changelog.txt
Version 0.9.9 Thursday 10th February 2005 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Re-release 0.9.8.11 as 0.9.9 stable

Version 0.9.8.11 Friday 14th January 2005 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Fixed use of split() function that broke Python 1.5.2 compatibility in config_affairs.py
- Added traps for ValueError in proxy_client.py for Python 1.5.2 compatibility
- Split testing of upstream proxy codepath in monitor_upstream.py into Python 2.x and 1.5.2 branches

Version 0.9.8.10 Friday 24th December Darryl Dixon <esrever_otua@users.sourceforge.net>
- Caught another exception condition for failed upstream proxy in monitor_upstream.py
- Reworked program flow in monitor_upstream.py and server.py to fix 100% cpu bug introduced in 0.9.8.9

Version 0.9.8.9 Thursday 2nd December Darryl Dixon <esrever_otua@users.sourceforge.net>
- Change default values for PARENT_PROXY_PORT to 8080 and PARENT_PROXY_TIMEOUT to 15 in server.cfg
- Squashed a bug in config_affairs.py that broke LISTEN_PORT when no PARENT_PROXY specified 
- Caught another exception condition in proxy_client.py
- Clarified 'Could not create socket' error in server.py to reference port number
- Added clause in monitor_upstream.py to prevent automatic failover in the case of single upstream proxy
- Removed redundant 'pass' statement from server.py
- Caught another exception condition for failed upstream proxy in monitor_upstream.py
- BUGFIX sourceforge ID#1070604: Added exception handler for call to socket.gethostbyname_ex() in config_affairs.py
- Moved copyright banner to appear earlier in initialisation
- Added TODO to monitor_upstream.py
- BUGFIX sourceforge ID#1070605: Added exception handler for call to open() in http_header.py

Version 0.9.8.8 Tuesday 16th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Squashed bug introduced in 0.9.8.7 with bad logger entry in proxy_client.py
- Cleaned up handling of 'PASSWORD' config item in config_affairs.py to squash potential bug
- Modified server.cfg to enhance description of 'PASSWORD' and 'USER' items
- Slightly updated the readme.txt

Version 0.9.8.7 Friday 12th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- BUGFIX sourceforge ID#798824: Modified exception trap for  c_method = self.client_head_obj.get_http_method() in proxy_client.py
- Replaced several TODOs from proxy_client.py with logger entries
- Removed bogus 'accepting connections' comment from main.py
- Squashed bug with SOMAXCONN setup trapping wrong exception
- Moved some initialisation code from main.py to server.py
- Clarified some code dealing with configuration items in main.py
- Modified some of the startup text/order
- Sanified processing of int() items in config_affairs.py
- Removed redundant statement from NTLM section of config_affairs.py
- Modified ntlm_auth.py to reflect NTLM_TO_BASIC as already an int
- Properly corrected second instance of sourceforge ID#1061067 in ntlm_auth.py

Version 0.9.8.6 Thursday 11th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Added MAX_CONNECTION_BACKLOG configuration item
- FEATURE/BUGFIX sourceforge request ID#831606: MAX_CONNECTION_BACKLOG now allows configurable listener
- Clarified some historical entries in changelog
- Minor cleanups of error text in config_affairs.py

Version 0.9.8.5 Wednesday 10th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Added PARENT_PROXY_TIMEOUT configuration item
- Squashed a typo bug in config_affairs.py for HOST_IP_LIST
- Updated Install.txt
- BUGFIX sourceforge request ID#707725 and ID#1061067: Fixed 2 instances of error in ntlm_auth.py

Version 0.9.8.4 Tuesday 9th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Changed method of delivering signal to just call sigHandler() directly
- Seriously reworked programme flow of monitor_upstream.py to accomodate change above
- Added locking of critical sections to keep threads sane
- Caught another exception in proxy_client.py

Version 0.9.8.3 Sunday 7th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Changed SIGALRM timer to new custom timerThread() (more portable)
- Cleaned up signal handling to accomodate new timer implementation
- Caught some more exceptions in proxy_client.py
- Fixed corner case which b0rked timer

Version 0.9.8.2 Friday 5th November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Added extra exception trapping for fault conditions in proxy_client.py
- Added __init__.py to begin making ntlmaps relocatable
- Changed from SIGUSR1 to SIGINT to make more portable
- Removed two non-ascii comments causing warnings from compiler in md4.py and des_c.py

Version 0.9.8.1 Monday 1st November 2004 Darryl Dixon <esrever_otua@users.sourceforge.net>
- Added monitor_upstream.py for upstream mode that detects upstream proxy failure
- Added ability to specify multiple upstream proxies for failover

Verison 0.9.8 - Fri May 17 04:07:20 MSD 2002
- internal redesign
- config file redesign
- fixed bug during HTTPS CONNECT authentication.
- fixed bug with UNICODE string conversion in NTLM msg3 creation code.
- no need in proxy port when proxy is not used
- fixed minor bug with an exception that was eraised if there was no http header
  in server's response.
- MSN Messenger and clients alike work again. It had been broken since APS 0.9.1
- minor bug in header remake (Proxy Connection -> Connection)
- fixed bug when client sends its header slowly and clients thread exits before
  doing anything useful. This was broken since version 0.9.7
- new optional value in config file NT_HOSTNAME (see comment in server.cfg).
- DOMAIN value in config is now NT_DOMAIN, to make it clear what domain name has
  to be used.
- implemented NTLM to BASIC translation. (it gives only one try to enter credentials now)

Verison 0.9.7 - Tue Dec 11 10:43:58 MSK 2001

main.py
#! /usr/bin/python

# This file is part of 'NTLM Authorization Proxy Server'
# Copyright 2001 Dmitry A. Rozmanov <dima@xenon.spb.ru>
#
# NTLM APS is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# NTLM APS is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the sofware; see the file COPYING. If not, write to the
# Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
#
import __init__

import sys

import server, config, config_affairs


#--------------------------------------------------------------
# config affairs
# look for default config name in lib/config.py
conf = config.read_config(config.findConfigFileNameInArgv(sys.argv, __init__.ntlmaps_dir+'/'))

conf['GENERAL']['VERSION'] = '0.9.9'

#--------------------------------------------------------------
print 'NTLM authorization Proxy Server v%s' % conf['GENERAL']['VERSION']
print 'Copyright (C) 2001-2004 by Dmitry Rozmanov and others.'

config = config_affairs.arrange(conf)


#--------------------------------------------------------------
# let's run it
serv = server.AuthProxyServer(config)
serv.run()


config_affairs.py
# This file is part of 'NTLM Authorization Proxy Server'
# Copyright 2001 Dmitry A. Rozmanov <dima@xenon.spb.ru>
#
# NTLM APS is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# NTLM APS is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the sofware; see the file COPYING. If not, write to the
# Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
#

import socket, thread, string, sys
import logger

#-------------------------------------------------------------------------
def arrange(conf):
    ""

    #-----------------------------------------------
    # GENERAL
    conf['GENERAL']['PARENT_PROXY']

    # if we do not use proxy then we do not need its port
    if conf['GENERAL']['PARENT_PROXY']:
        conf['GENERAL']['AVAILABLE_PROXY_LIST'] = string.split(conf['GENERAL']['PARENT_PROXY'])
        conf['GENERAL']['PARENT_PROXY'] = conf['GENERAL']['AVAILABLE_PROXY_LIST'].pop()
        conf['GENERAL']['PARENT_PROXY_PORT'] = makeInt(conf['GENERAL']['PARENT_PROXY_PORT'], 'PARENT_PROXY_PORT')
        conf['GENERAL']['PARENT_PROXY_TIMEOUT'] = makeInt(conf['GENERAL']['PARENT_PROXY_TIMEOUT'], 'PARENT_PROXY_TIMEOUT')
    try:
        conf['GENERAL']['MAX_CONNECTION_BACKLOG'] = int(conf['GENERAL']['MAX_CONNECTION_BACKLOG'])
    except ValueError:
        if conf['GENERAL']['MAX_CONNECTION_BACKLOG'] == 'SOMAXCONN':
            conf['GENERAL']['MAX_CONNECTION_BACKLOG'] = socket.SOMAXCONN
        else:
            print "ERROR: There is a problem with 'MAX_CONNECTION_BACKLOG' in the config (neither a number nor 'SOMAXCONN'?)"
            sys.exit(1)
    conf['GENERAL']['LISTEN_PORT'] = makeInt(conf['GENERAL']['LISTEN_PORT'], 'LISTEN_PORT')

    conf['GENERAL']['ALLOW_EXTERNAL_CLIENTS'] = makeInt(conf['GENERAL']['ALLOW_EXTERNAL_CLIENTS'], 'ALLOW_EXTERNAL_CLIENTS')
    hostname = socket.gethostname()
    conf['GENERAL']['HOST'] = hostname
    try:
        externalIP = socket.gethostbyname_ex(hostname)[2]
    except socket.gaierror:
        print "ERROR: Unable to get the IP address of this machine.  This is not a fatal problem, but may cause problems for you using this proxy in some scenarios."
        externalIP = []
    conf['GENERAL']['HOST_IP_LIST'] = externalIP + ['127.0.0.1']

    conf['GENERAL']['FRIENDLY_IPS'] = conf['GENERAL']['HOST_IP_LIST'] + string.split(conf['GENERAL']['FRIENDLY_IPS'])

    conf['GENERAL']['URL_LOG'] = makeInt(conf['GENERAL']['URL_LOG'], 'URL_LOG')
    url_logger = logger.Logger('url.log', conf['GENERAL']['URL_LOG'])
    url_logger_lock = thread.allocate_lock()
    conf['GENERAL']['URL_LOGGER'] = url_logger
    conf['GENERAL']['URL_LOG_LOCK'] = url_logger_lock


    #-----------------------------------------------
    # NTLM_AUTH
    if not conf['NTLM_AUTH'].has_key('NTLM_FLAGS'):
        conf['NTLM_AUTH']['NTLM_FLAGS'] = ''
    #conf['NTLM']['FULL_NTLM'] = makeInt(conf['NTLM']['FULL_NTLM'], 'FULL_NTLM')
    conf['NTLM_AUTH']['LM_PART'] = makeInt(conf['NTLM_AUTH']['LM_PART'], 'LM_PART')
    conf['NTLM_AUTH']['NT_PART'] = makeInt(conf['NTLM_AUTH']['NT_PART'], 'NT_PART')
    conf['NTLM_AUTH']['NTLM_TO_BASIC'] = makeInt(conf['NTLM_AUTH']['NTLM_TO_BASIC'], 'NTLM_TO_BASIC')
    if not conf['NTLM_AUTH']['NT_DOMAIN']:
        print "ERROR: NT DOMAIN must be set."
        sys.exit(1)
    if not conf['NTLM_AUTH'].has_key('PASSWORD'):
        conf['NTLM_AUTH']['PASSWORD'] = ''


    #-----------------------------------------------
    # DEBUG
    conf['DEBUG']['DEBUG'] = makeInt(conf['DEBUG']['DEBUG'], 'DEBUG')
    conf['DEBUG']['AUTH_DEBUG'] = makeInt(conf['DEBUG']['AUTH_DEBUG'], 'AUTH_DEBUG')
    conf['DEBUG']['BIN_DEBUG'] = makeInt(conf['DEBUG']['BIN_DEBUG'], 'BIN_DEBUG')

    # screen activity
    if conf['DEBUG'].has_key('SCR_DEBUG'):
           conf['DEBUG']['SCR_DEBUG'] = int(conf['DEBUG']['SCR_DEBUG'])
    else:
        conf['DEBUG']['SCR_DEBUG'] = 0

    return conf

def makeInt(string, errorDesc='an item'):
    try:
        ret = int(string)
    except ValueError:
        print "ERROR: There is a problem with "+errorDesc+" in the config (is it not a number?)"
        sys.exit(1)
    return ret
